### PR TITLE
perf(freespace_planning_algorithms): improve freespace planning performance

### PR DESCRIPTION
**Pull request by NorahXiong**
## Description

See #2039

## Tests performed

Test environment: planning simulator as default
- start point: "pose: {position: {x: 3713.99, y: 73732.0, z: 0.664812}, orientation: {x: 0.0, y: 0.0, z: 0.850511, w: 0.525957}"
- goal point: "pose: {position: {x: 3727.39, y: 73750.0, z: 0.0}, orientation: {x: 0.0, y: 0.0, z: 0.850592, w: 0.525827}"
- In test 1 costmap resolution is 0.2m and in test 2 which is 0.1m.
- Results are average value of at least 3 replications.

**Use unordered_map to store node data instead of vector**
The memory cost sharply decreases. The efficiency of data access is almost the same as before.
image

**Eliminate unnecessary calculation steps in collision detection function**
Delete duplicate indexes which corresponds to vehicle shape. More than 3/4 indexes are not unique.
image

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.
- [x] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/freespace_planning_algorithms/src/astar_search.cpp
Comment (maxime-clem) [inline planning/freespace_planning_algorithms/src/astar_search.cpp:143]: How did you choose the value `100000` ? Can it be calculated from the `costmap` dimensions and the `y_scale_` ?

Comment (NorahXiong) [reply planning/freespace_planning_algorithms/src/astar_search.cpp:143]: I did considered to set the value dynamically according to map parameters but the nodes actually needed varies largely in different cases even using the same map. I've tested different cases in this [discussion](https://github.com/orgs/autowarefoundation/discussions/2716#discussion-4207199). The "Visited-nodes" column corresponds to nodes quantity. The number could be larger in low-resolution map than in high-resolution map if the case using low-resolution map is complicated enough.

On the other hand, the initial scale of graph_ seems not really allocated in memory. So maybe it will not have an obvious influence on efficiency.

By the way, could you please review this PR?

**Review by maxime-clem — APPROVED** [review]
The change makes sense, the code looks good, and parking is working well in simulation.

File: planning/freespace_planning_algorithms/src/astar_search.cpp
Comment (maxime-clem) [reply planning/freespace_planning_algorithms/src/astar_search.cpp:143]: I am sorry the discussion you opened went unnoticed. The observations you made are very good and your PR makes a lot of sense. Once this is merged, you can post a link of the PR in the discussion and mark it as the answer.

In the future, if people do not react to your discussion, you can notify people during the software working group or on Discord.

Comment (NorahXiong) [reply planning/freespace_planning_algorithms/src/astar_search.cpp:143]: > I understand now, thank you. I think in that case it is better to not `reserve` here.

I compared settings with and without "graph_.reserve(n)". I think it's better to keep the reserve action cause it can make the performance more stable. Test results are as follow.

A complicated case: nearly 300,000 nodes are expanded.
- start point: "pose: {position: {x: 3713.99, y: 73732.0, z: 0.664812}, orientation: {x: 0.0, y: 0.0, z: 0.850511, w: 0.525957}"
- goal point: "pose: {position: {x: 3746.88, y: 73748.5, z: 0.0}, orientation: {x: 0.0, y: 0.0, z: 0.847164, w: 0.531332}"

image

image

From the table above, we can conclude:
- Without pre-allocation, it may cost more time due to reasons like frequent rehash.
- Smaller pre-allocation size has little influence on general performance. (There may exist a key size that is sensitive to performance, still need testing.)

Comment (NorahXiong) [reply planning/freespace_planning_algorithms/src/astar_search.cpp:143]: > I am sorry the discussion you opened went unnoticed. The observations you made are very good and your PR makes a lot of sense. Once this is merged, you can post a link of the PR in the discussion and mark it as the answer.
> 
> In the future, if people do not react to your discussion, you can notify people during the software working group or on Discord.

I don't know why the discussion link is not permanent. Thanks for your suggestion. I will paste this PR to the discussion later.

**Comment by TakaHoribe** [comment]
@NorahXiong Thank you for your contribution! Your report for the performance improvement is very impressive!

Recently, we merged a large PR for the freespace planner, so there is now a conflict in this PR. Would you solve the conflict? Then we'd like to merge this PR soon.

**Comment by liuXinGangChina** [comment]
@TakaHoribe ,hi. NorthXiong has resolve the conflict , please take a look and give your valuable advice.

**Comment by liuXinGangChina** [comment]
> @NorahXiong Thank you for your contribution! Your report for the performance improvement is very impressive!
> 
> Recently, we merged a large PR for the freespace planner, so there is now a conflict in this PR. Would you solve the conflict? Then we'd like to merge this PR soon.

@TakaHoribe  ,hi. NorthXiong has resolved the conflict , please take a look and give your valuable advice.

**Comment by maxime-clem** [comment]
@liuXinGangChina @NorahXiong One test is failing in the CI: https://github.com/autowarefoundation/autoware.universe/actions/runs/3376624609/jobs/5605343502#step:7:380
Please fix it as we cannot merge until all tests are passing.

**Comment by NorahXiong** [comment]
> @liuXinGangChina @NorahXiong One test is failing in the CI: https://github.com/autowarefoundation/autoware.universe/actions/runs/3376624609/jobs/5605343502#step:7:380 Please fix it as we cannot merge until all tests are passing.

@maxime-clem @TakaHoribe  I've fixed the problem. PR is ready to merge now.

**Comment by TakaHoribe** [comment]
@NorahXiong Thank you so much. I'd like to merge this PR soon. Do you have a test script to measure performance like you put in the top of this PR? We need to track the performance since the OSS is frequently updated. Your script would be very useful.

**Comment by NorahXiong** [comment]
> @NorahXiong Thank you so much. I'd like to merge this PR soon. Do you have a test script to measure performance like you put in the top of this PR? We need to track the performance since the OSS is frequently updated. Your script would be very useful.

Sorry I don't have that script. I counted executing time of each step by debug log. I have a simple script to precisely set the start and goal point in order to repeat tests. I'd like to share it if you need.

**Comment by NorahXiong** [comment]
> > @NorahXiong Thank you so much. I'd like to merge this PR soon. Do you have a test script to measure performance like you put in the top of this PR? We need to track the performance since the OSS is frequently updated. Your script would be very useful.
> 
> Sorry I don't have that script. I counted executing time of each step by debug log. I have a simple script to precisely set the start and goal point in order to repeat tests. I'd like to share it if you need.

@TakaHoribe, the attached script can precisely set start/goal point and the usage is as below. Hope it will be helpful for you.
```
# source env
./astar_set 1 n # set start point with the nth start point
./astar_set 2 n # set start point with the nth goal point
```

[astar_set.zip](https://github.com/autowarefoundation/autoware.universe/files/10020746/astar_set.zip)

**Review by TakaHoribe — APPROVED** [review]
@NorahXiong Thank you. I could confirm the great improvement of the freespace planner.

**Pull request merged**
